src/ser/variant_ser.rs

```diff
@@ -0,0 +1,315 @@
+        Err(VariantStructSerializerError::ExpectedStruct)
+    }
+
+    fn serialize_unit_variant(
+        self,
+        _name: &'static str,
+        _variant_index: u32,
+        _variant: &'static str,
+    ) -> Result<Self::Ok, Self::Error> {
+        Err(VariantStructSerializerError::ExpectedStruct)
+    }
+
+    fn serialize_newtype_struct<T>(
+        self,
+        _name: &'static str,
+        _value: &T,
+    ) -> Result<Self::Ok, Self::Error>
+    where
+        T: ?Sized + Serialize,
+    {
+        Err(VariantStructSerializerError::ExpectedStruct)
+    }
+
+    fn serialize_newtype_variant<T>(
+        self,
+        _name: &'static str,
+        _variant_index: u32,
+        _variant: &'static str,
+        _value: &T,
+    ) -> Result<Self::Ok, Self::Error>
+    where
+        T: ?Sized + Serialize,
+    {
+        Err(VariantStructSerializerError::ExpectedStruct)
+    }
+
+    fn serialize_seq(self, _len: Option<usize>) -> Result<Self::SerializeSeq, Self::Error> {
+        Err(VariantStructSerializerError::ExpectedStruct)
+    }
+
+    fn serialize_tuple(self, _len: usize) -> Result<Self::SerializeTuple, Self::Error> {
+        Err(VariantStructSerializerError::ExpectedStruct)
+    }
+
+    fn serialize_tuple_struct(
+        self,
+        _name: &'static str,
+        _len: usize,
+    ) -> Result<Self::SerializeTupleStruct, Self::Error> {
+        Err(VariantStructSerializerError::ExpectedStruct)
+    }
+
+    fn serialize_tuple_variant(
+        self,
+        _name: &'static str,
+        _variant_index: u32,
+        _variant: &'static str,
+        _len: usize,
+    ) -> Result<Self::SerializeTupleVariant, Self::Error> {
+        Err(VariantStructSerializerError::ExpectedStruct)
+    }
+
+    fn serialize_map(self, _len: Option<usize>) -> Result<Self::SerializeMap, Self::Error> {
+        Err(VariantStructSerializerError::ExpectedStruct)
+    }
+
+    fn serialize_struct_variant(
+        self,
+        _name: &'static str,
+        _variant_index: u32,
+        _variant: &'static str,
+        _len: usize,
+    ) -> Result<Self::SerializeStructVariant, Self::Error> {
+        Err(VariantStructSerializerError::ExpectedStruct)
+    }
+}
+
+macro_rules! match_type_unsafe {
+    ($target_type:ty, $value:ident) => {
+        || {
+            if typeid::of::<T>() == typeid::of::<$target_type>() {
+                Some(Variant::from(*unsafe {
+                    std::mem::transmute_copy::<&T, &$target_type>(&$value)
+                }))
+            } else {
+                None
+            }
+        }
+    };
+}
+
+impl SerializeStruct for VariantStructSerializer {
+    type Ok = <VariantStructSerializer as Serializer>::Ok;
+
+    type Error = VariantStructSerializerError;
+
+    fn serialize_field<T>(&mut self, key: &'static str, value: &T) -> Result<(), Self::Error>
+    where
+        T: ?Sized + Serialize,
+    {
+        let variant = match_type_unsafe!((), value)()
```

Comment:
This is the "awful serialization hack" mentioned in the PR description.
I could find absolutely no way to implement a generic Variant conversion that fails by default but works when explicitly defined (it could be done with auto traits or specialization, but neither is stable Rust), and so this is all I could think of for a solution.

Any better ideas would be greatly appreciated.